Exclude interface's primary address from IP pool by default in Azure

hemanthmalla opened the pull request:
Currently `--azure-use-primary-address` defaults to true and this can lead
to connectivity issues in host networked pods unless users explicitly
disable DHCP on the interface.

See https://github.com/cilium/cilium/pull/13415 for more details on connectivity issues. 
Defaulting to false also makes the behavior consistent with ENI mode.

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

Open questions  : 
- [ ] How do i make sure this change shows up in upgrade guides ?
- [ ] What happens to existing pods using the primary IP for users upgrading from prior versions ?
- [ ] Is this too late for `1.12` ?

hemanthmalla committed:
Exclude interface's primary address from IP pool by default in Azure

Currently --azure-use-primary-address defaults to true and this can lead
to connectivity issues in host networked pods unless users explicitly
disable DHCP on the interface.

See https://github.com/cilium/cilium/pull/13415 for more details on
connectivity issues. Defaulting to false also makes the behavior
consistent with ENI mode.

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

christarazi reviewed (changes requested):
Change makes sense given that it's converging with the ENI implementation.

1) Please add a note in the [upgrade guide](https://docs.cilium.io/en/latest/operations/upgrade/#upgrade-notes).
2) I'm not sure, but it sounds like users will need to restart the workload pods which have those IPs so that Cilium can assign new ones that are non-primary. I'm not sure it's worth coming up with a solution to help make this a hitless upgrade, but we can discuss this in the community meeting if you'd like.
3) Let's also bring that up in the community meeting

hemanthmalla committed:
Add support for bypassing ip availability upon restore in Azure IPAM

We need support for --bypass-ip-availability-upon-restore in Azure IPAM
mode because 1.12 switches the default for --azure-use-primary-address
from true to false. Users upgrading from prior versions might already be
using interface's primary address for pods. Without this flag pods using
primary addresses cannot get a new IP addresses.

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

hemanthmalla committed:
Support backward compatibility for --azure-use-primary-address in helm

Signed-off-by: Hemanth Malla <hemanth.malla@datadoghq.com>

gandro reviewed (approved):
Thanks!
